util/lgrind/source/lgrindef.5
.\" @(#)lgrindef.5 3.62 11/2/97 MPi; from UCB 4.3
.\" Copyright (c) 1983 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\" $Id: lgrindef.5,v 1.4 1999/10/05 17:41:35 mike Exp $
.\"
.\" This is the standard SunOS vgrindefs(5) manpage, with a few
.\" additions to describe the extra capabilities needed by lgrind,
.\" viz.: zb, ze, tb, te, mb, me, vb, ve, cf, rb.
.\"
.tr ||
.TH LGRINDEF 5 "11 February 1997" "TeX/LaTeX"
.SH NAME
lgrindef \- LGrind's language definition data base
.SH NOTE
This man page is not yet much outdated, but might be soon except somebody
asks me to work on it. Consider the LaTeX docs the real docs.
.SH SYNOPSIS
.B /usr/TeX/texmf/tex/latex/lgrind/lgrindef
.SH DESCRIPTION
.IX  "lgrindef file"  ""  "\fLlgrindef\fP \(em LGrind language definitions"
.B lgrindef
contains all language definitions for
.BR lgrind (1).
The data base is very similar to
.BR vgrind (5)
and
.BR termcap (5),
and it is upward-compatible with that of
.BR vgrind (5).
Capabilities in
.B lgrindef
are of two types:
Boolean capabilities which indicate that the language has
some particular feature
and string
capabilities which give a regular expression or
keyword list.
Entries may continue onto multiple lines by giving a \e as the last
character of a line.
Lines starting with # are comments.
.SS Capabilities
The following table names and describes each capability.
.PP
.PD 0
.ta \w'\fBName\fP  'u
.nr Xx \w'\fBName\fP  \fBType\fP  'u
.TP \n(Xxu
.B
Name	Type
.B Description
.TP
.B ab	\fRstr
Regular expression for the start of an alternate form comment
.TP
.B ae	\fRstr
Regular expression for the end of an alternate form comment
.TP
.B bb	\fRstr
Regular expression for the start of a block
.TP
.B be	\fRstr
Regular expression for the end of a lexical block
.TP
.B cb	\fRstr
Regular expression for the start of a comment
.TP
.B ce	\fRstr
Regular expression for the end of a comment
.TP
.B cf	\fRbool
(Boolean) Use specialized C function detection
.TP 
.B id	\fRstr
String giving characters other than letters and digits
that may legally occur in identifiers (default `_')
.TP
.B kw	\fRstr
A list of keywords separated by spaces
.TP
.B lb	\fRstr
Regular expression for the start of a character constant
.TP
.B le	\fRstr
Regular expression for the end of a character constant
.TP
.B mb	\fRstr
Regular expression for the start of TeX math within a comment
.TP
.B me	\fRstr
Regular expression for the end of TeX math within a comment
.TP
.B np	\fRstr
Regular expression for a line
.B not
containing the start of a procedure
.TP
.B oc	\fRbool
Present means upper and lower case are equivalent
.TP
.B pb	\fRstr
Regular expression for start of a procedure
.TP
.B pl	\fRbool
Procedure definitions are constrained to the lexical level
matched by the `px' capability
.TP
.B px	\fRstr
A match for this regular expression indicates
that procedure definitions may occur at the next lexical level.
Useful for lisp-like languages in which procedure definitions
occur as subexpressions of defuns.
.TP
.B rb	\fRstr
Regular expression for the start of a block outside the actual code
.TP
.B sb	\fRstr
Regular expression for the start of a string
.TP
.B se	\fRstr
Regular expression for the end of a string
.TP
.B rb	\fRstr
Regular expression for the end of a block outside a funtion
(e. g. records in Pascal and Modula-2)
.TP
.B tb	\fRstr
Regular expression for the start of TeX text within a comment
.TP
.B tc	\fRstr
Use the named entry as a continuation of this one
.TP
.B te	\fRstr
Regular expression for the end of TeX text within a comment
.TP
.B tl	\fRbool
Present means procedures are only defined at the top lexical level
.TP
.B vb	\fRstr
Regular expression for the start of typewriter text within a comment
.TP
.B ve	\fRstr
Regular expression for the end of typewriter text within a comment
.TP
.B zb	\fRstr
Regular expression for the start of program text within a comment
.TP
.B ze	\fRstr
Regular expression for the end of program text within a comment
.PD
.DT
.SS "Regular Expressions"
.B lgrindef
uses regular expressions similar to those of
.BR ex (1)
and
.BR lex (1).
The characters `^',
`$',
`\^|',
`:',
and `\e'
are reserved characters and must be
`quoted' with a preceding \e if they
are to be included as normal characters.
The metasymbols and their meanings are:
.IP $
The end of a line
.PD 0.2v
.IP ^
The beginning of a line
.IP \ed
A delimiter (space, tab, newline, start of line)
.IP \ea
Matches any string of symbols
(like `.*' in lex)
.IP \ep
Matches any identifier.
In a procedure definition
(the `pb' capability)
the string that matches this symbol is used as the procedure name.
.IP (\^)
Grouping
.IP |
Alternation
.IP ?
Last item is optional
.IP \ee
Preceding any string means that the string will not match an
input string if the input string is preceded by an escape character (\e).
This is typically used for languages (like C) that can include the
string delimiter in a string by escaping it.
.PD
.PP
Unlike other regular expressions in the system,
these match words and not characters.
Hence something like `(tramp\^|\^steamer)flies?'
would match `tramp',
`steamer',
`trampflies',
or `steamerflies'.
Contrary to some forms of regular expressions,
.B lgrindef
alternation binds very tightly.
Grouping parentheses are likely to be necessary in expressions
involving alternation.
.SS "Keyword List"
The keyword list is just a list of keywords in the language separated
by spaces.
If the `oc' boolean is specified,
indicating that upper and lower case are equivalent,
then all the keywords should be specified in lower case.
.SH EXAMPLE
The following entry,
which describes the C language,
is typical of a language entry.
.IP
.ft B
.nf
C\^|\^the C programming language:\e
	:pb=^\ed?*?\ed?\ep\ed?\(\ea?\):bb={:be=}:cb=/*:ce=*/:\e
	:sb=":se=\ee":lb=':le=\ee':tl:\e
	:zb=@:ze=@:tb=%%:te=%%:mb=%\e$:me=\e$%:vb=%\e\^|:ve=\e\^|%:\e
	:kw=asm auto break case char continue default do double\e
	else enum extern float for fortran goto if int long\e
	register return short sizeof static struct switch typedef\e
	union unsigned while #define #else #endif #if #ifdef\e
	#ifndef #include #undef # define else endif if ifdef\e
	ifndef include undef:
.fi
.ft
.PP
Note that the first field is just the language name
(and any variants of it).
Thus the C language could be specified to
.BR lgrind (1)
as `c' or `C', since case is not significant here.
.SH FILES
.ta \w'/usr/TeX/texmf/tex/latex/lgrind/lgrindef   'u
\fB/usr/TeX/texmf/tex/latex/lgrind/lgrindef\fR	file containing terminal descriptions
.SH "SEE ALSO"
.BR latex (1),
.BR lgrind (1),
.BR vgrindefs (5),
For full documentation, refer to the package itself; it comes as a .dtx
containing both the documentation and the LaTeX-files.



util/lgrind/source/lgrindef.c
#ifndef lint
static char sccsid[] = "@(#)lgrindef.c   4.3 (Berkeley) 11/12/84";
static char rcsid[] =
   "$Id: lgrindef.c,v 1.3 1999/05/28 11:00:16 mike Exp $";
#endif

/* Copyright %%\copyright%% 1979 Regents of the University of California 
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#define MAXHOP	32	/* max number of %|tc=|% indirections */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include "lgrindef.h"

/*
 * grindcap --- routines for dealing with the language definitions data base
 *	(code stolen almost totally from termcap)
 *
 * BUG:		Should use a "last" pointer in @tbuf@, so that searching
 *		for capabilities alphabetically would not be a %$n^2/2$%
 *		process when large numbers of capabilities are given.
 * Note:	If we add a last pointer now we will screw up the
 *		%|tc|% capability. We really should compile termcap.
 *
 * Essentially all the work here is scanning and decoding escapes
 * in string capabilities.  We don't use stdio because the editor
 * doesn't, and because living w/o it is not hard.
 */

static	char *tbuf;
static	char *filename;
static	int hopcount = 0; /* detect infinite loops in termcap */

static char	*tskip(char *);
static char	*tdecode(char *, char **);

/*
 * @tnchktc@: check the last entry, see if it's %|tc=xxx|%. If so,
 * recursively find %|xxx|% and append that entry (minus the names)
 * to take the place of the %|tc=xxx|% entry. This allows termcap
 * entries to say "like an HP2621 but doesn't turn on the labels".
 * Note that this works because of the left to right scan.
 */
static int tnchktc(void)
{
   register char *p, *q;
   char tcname[16];	/* name of similar terminal */
   char tcbuf[BUFFERSIZE];
   char *holdtbuf = tbuf;
   int l;
   
   p = tbuf + strlen(tbuf) - 2;	/* before the last colon */
   while (*--p != ':')
      if (p < tbuf) {
     /* Not very informative. Wrong BTW (for the configs). Shut up. */ 
     /* write(2, "Bad lgrind entry\n", 18); */
	 return 0;
      }
   p++;
   /* @p@ now points to beginning of last field */
   if (p[0] != 't' || p[1] != 'c')
      return 1;
   strcpy(tcname,p+3);
   q = tcname;
   while (q && *q != ':')
      q++;
   *q = '\0';
   if (++hopcount > MAXHOP) {
      fprintf(stderr, "Infinite tc= loop\n");
      return 0;
   }
   if (tgetent(tcbuf, tcname, filename) != 1)
      return 0;
   for (q = tcbuf; *q != ':'; q++)
      ;
   l = (int)(p - holdtbuf + strlen(q));
   if (l > BUFFERSIZE) {
      fprintf(stderr, "LGrind entry too long\n");
      q[(int)(BUFFERSIZE - (p-tbuf))] = '\0';
   }
   strcpy(p, q+1);
   tbuf = holdtbuf;
   return 1;
}


/*
 * @tnamatch@ deals with name matching.  The first field of the termcap
 * entry is a sequence of names separated by %|||%'s, so we compare
 * against each such name.  The normal %|:|% terminator after the last
 * name (before the first field) stops us.
 */
static int tnamatch(char *np)
{
   register char *Np, *Bp;
   
   Bp = tbuf;
   if (*Bp == '#')
      return 0;
   for (;;) {
      for (Np = np; *Np && toupper(*Bp) == toupper(*Np); Bp++, Np++)
		 continue;
      if (*Np == '\0' && (*Bp == '|' || *Bp == ':' || *Bp == '\0'))
		 return 1;
      while (*Bp != '\0' && *Bp != ':' && *Bp != '|')
		 Bp++;
      if (*Bp == '\0' || *Bp == ':')
		 return 0;
      Bp++;
   }
}


/*
 * Skip to the next field.  Notice that this is very dumb, not
 * knowing about %|\:|% escapes or any such.  If necessary, %|:|%'s can
 * be put into the termcap file in octal.
 */
static char *tskip(register char *bp)
{
   while (*bp && *bp != ':')
      bp++;
   if (*bp == ':')
      bp++;
   return bp;
}


/*
 * Return the (numeric) option id.
 * Numeric options look like
 *	%|li#80|%
 * i.e. the option string is separated from the numeric value by
 * a %|#|% character.  If the option is not found we return %$-1$%.
 * Note that we handle octal numbers beginning with %$0$%.
 */
int tgetnum(char *id)
{
   register int i, base;
   register char *bp = tbuf;
   
   for (;;) {
      bp = tskip(bp);
      if (*bp == '\0')
	 return -1;
      if (*bp++ != id[0] || *bp == '\0' || *bp++ != id[1])
	 continue;
      if (*bp == '@')
	 return -1;
      if (*bp != '#')
	 continue;
      bp++;
      base = 10;
      if (*bp == '0')
	 base = 8;
      i = 0;
      while (isdigit(*bp))
	 i *= base, i += *bp++ - '0';
      return i;
   }
}



/*
 * Handle a flag option.
 * Flag options are given "naked", i.e. followed by a %|:|% or the end
 * of the buffer.  Return 1 if we find the option, or 0 if it is
 * not given.
 */
int tgetflag(char *id)
{
   register char *bp = tbuf;
   
   for (;;) {
      bp = tskip(bp);
      if (!*bp)
	 return 0;
      if (*bp++ == id[0] && *bp != 0 && *bp++ == id[1]) {
	 if (!*bp || *bp == ':')
	    return 1;
	 else if (*bp == '@')
	    return 0;
      }
   }
}


/*
 * Get a string valued option.
 * These are given as
 *	%|cl=^Z|%
 * Much decoding is done on the strings, and the strings are
 * placed in area, which is a ref parameter which is updated.
 * No checking on area overflow.
 */
char *tgetstr(char *id, char **area)
{
	register char *bp = tbuf, *i;
	register int eq;

	for (;;) {
		bp = tskip(bp);
		if (!*bp)
			return 0;
		eq=1;
		for (i=id; *i;)
			if (*i++!=*bp++) { eq=0; break; }
		if (!eq)
			continue;
		if (*bp == '@')
			return 0;
		if (*bp != '=')
			continue;
		bp++;
		return tdecode(bp, area);
	}
}


/*
 * @tdecode@ does the grunge work to decode the
 * string capability escapes.
 */
static char *tdecode(register char *str, char **area)
{
   char *cp = *area, c;
   
   while ((c = *str++) != 0) {
      if (c == ':') {
         if (*(cp-1) == '\\')
            cp--;
         else           
            break;
      }
      *cp++ = c;
   }
   *cp++ = '\0';
   str = *area;
   *area = cp;
   return str;
}

/*
 * Get an entry for terminal name in buffer @bp@,
 * from the termcap file.  Parse is very rudimentary;
 * we just notice escaped newlines.
 */
int tgetent(char *bp, char *name, char *file)
{
   register char *cp;
   register char c;        /* 30.7.96 -- tryal [sic] */
   register int i = 0, cnt = 0;
   char ibuf[BUFFERSIZE];
   FILE *tf;
   
   tbuf = bp;
   tf = 0;
   filename = file;
   tf = fopen(filename, "rt");
   if (tf == NULL)
      return -1;
   for (;;) {
      cp = bp;
      for (;;) {
	 if (i == cnt) {
	    cnt = fread(ibuf, 1, BUFFERSIZE, tf);
	    if (cnt <= 0) {
	       fclose(tf);
	       return 0;
		}
		i = 0;
	 }
	 c = ibuf[i++];
	 if (c == '\n') {
		if (cp > bp && cp[-1] == '\\'){
		   cp--;
		   continue;
		}
		break;
	 }
	 if (cp >= bp+BUFFERSIZE) {
		fprintf(stderr, "LGrind entry too long\n");
		break;
	 } else
		*cp++ = c;
	  }
	  *cp = '\0';

	  /*
	   * The real work for the match.
	   */
	  if (tnamatch(name)) {
	 fclose(tf);
	 return tnchktc();
	  }
   }
}
